File: app/src/main/java/com/doxigo/muchtoman/Format.kt
package com.doxigo.muchtoman

import java.math.BigDecimal
import java.math.RoundingMode
import java.text.NumberFormat
import java.util.Locale
import kotlin.math.abs
import kotlin.math.roundToLong
import kotlin.math.truncate

private val FA: Locale = Locale.forLanguageTag("fa-IR")
private val FA_INTEGER = object : ThreadLocal<NumberFormat>() {
    override fun initialValue(): NumberFormat = NumberFormat.getIntegerInstance(FA)
}

/** Grouped Persian digits: 118500 -> "۱۱۸٬۵۰۰" */
fun faNumber(value: Double): String =
    FA_INTEGER.get()!!.format(value.roundToLong())

/**
 * Persian digits with decimals kept: 0.0345 -> "۰٫۰۳۴۵". With [pad] the decimals are held at
 * exactly [decimals] places, zeros and all, so a column of figures keeps one shape.
 */
fun faDecimal(value: Double, decimals: Int, pad: Boolean = false): String {
    val f = NumberFormat.getNumberInstance(FA)
    f.maximumFractionDigits = decimals
    f.minimumFractionDigits = if (pad) decimals else 0
    return f.format(value)
}

/**
 * Toman amounts get long fast (3000 USD is a nine-digit number of Toman), and a wall of
 * digits is exactly what an older reader can't scan. Iranians say the magnitude out loud,
 * so we do too: "۳۰۰ میلیون" instead of "۳۰۰٬۰۰۰٬۰۰۰".
 */
fun tomanOf(rial: Long): Double = rial / 10.0

fun faCompact(toman: Double, dec: Int = 1, pad: Boolean = false): String {
    val n = abs(toman)
    val (unit, div) = when {
        n >= 1_000_000_000_000.0 -> "همت" to 1_000_000_000_000.0   // هزار میلیارد
        n >= 1_000_000_000.0 -> "میلیارد" to 1_000_000_000.0
        n >= 1_000_000.0 -> "میلیون" to 1_000_000.0
        n >= 1_000.0 -> "هزار" to 1_000.0
        else -> return faNumber(toman)
    }
    // Truncated, never rounded. Rounding is what made 10.8 million read as "۱۱ میلیون", and
    // rounding to one decimal still turns 2.95 into "۳". For money there is only one safe
    // direction: never show more than she has. The exact figure sits alongside.
    //
    // Two forms, and the difference is deliberate. Prose (dec = 1, unpadded) drops a trailing
    // zero, because "۱۱ میلیون" is how the amount is actually said mid-sentence. The figures
    // on the list and the hero use dec = 3 padded: a holding is usually a whole number times a
    // round rate, so without padding they collapse to "۸۵۰٫۸" and the precision looks like it
    // comes and goes — except a figure that lands exactly on the unit, which drops its
    // decimals entirely: "۱۸۰٫۰۰۰ میلیون" is three zeros carrying nothing, and reads worse
    // than "۱۸۰ میلیون".
    var f = 1.0
    repeat(dec) { f *= 10 }
    val t = truncate(toman / div * f) / f
    return "${faDecimal(t, dec, pad && t % 1.0 != 0.0)} $unit"
}

private val ONES = arrayOf("", "یک", "دو", "سه", "چهار", "پنج", "شش", "هفت", "هشت", "نه")
private val TEENS = arrayOf(
    "ده", "یازده", "دوازده", "سیزده", "چهارده",
    "پانزده", "شانزده", "هفده", "هجده", "نوزده",
)
private val TENS = arrayOf("", "", "بیست", "سی", "چهل", "پنجاه", "شصت", "هفتاد", "هشتاد", "نود")
private val HUNDREDS =
    arrayOf("", "صد", "دویست", "سیصد", "چهارصد", "پانصد", "ششصد", "هفتصد", "هشتصد", "نهصد")
private val SCALES = arrayOf("", "هزار", "میلیون", "میلیارد", "هزار میلیارد")

/** 1..999 in words. */
private fun tripleToWords(n: Int): String {
    val parts = mutableListOf<String>()
    val h = n / 100
    val r = n % 100
    if (h > 0) parts += HUNDREDS[h]
    when {
        r in 10..19 -> parts += TEENS[r - 10]
        else -> {
            if (r / 10 > 0) parts += TENS[r / 10]
            if (r % 10 > 0) parts += ONES[r % 10]
        }
    }
    return parts.joinToString(" و ")
}

/**
 * The amount spelled out: 10_800_000 -> "ده میلیون و هشتصد هزار". Digits are quick to scan
 * but easy to misread by a factor of ten; the words are the check on that, which is the whole
 * reason they are here. Exact, never rounded.
 */
fun faWords(value: Long): String {
    if (value == 0L) return "صفر"
    if (value < 0) return "منفی ${faWords(-value)}"

    val groups = mutableListOf<Int>()      // least significant first
    var v = value
    while (v > 0) {
        groups += (v % 1000).toInt()
        v /= 1000
    }
    if (groups.size > SCALES.size) return "" // beyond هزار میلیارد; caller shows digits only

    val parts = mutableListOf<String>()
    for (i in groups.indices.reversed()) {
        val g = groups[i]
        if (g == 0) continue
        // "هزار" alone reads naturally; "یک هزار" does not. Every larger scale keeps its یک.
        val words = if (g == 1 && i == 1) "" else tripleToWords(g)
        parts += listOf(words, SCALES[i]).filter { it.isNotBlank() }.joinToString(" ")
    }
    return parts.joinToString(" و ")
}

/** Words for a Toman figure, or null when spelling it out would not help. */
fun faWordsToman(value: Double): String? {
    if (value <= 0 || value >= 1e15) return null
    val words = faWords(value.roundToLong())
    return if (words.isBlank()) null else "$words تومان"
}

/**
 * Reads a number the way a Persian keyboard actually produces it: Persian (۰-۹) or
 * Arabic-Indic (٠-٩) digits, Persian decimal separator, thousands separators ignored.
 * Returns null on anything it doesn't understand rather than guessing — this is the
 * money path, a silent misparse would show her the wrong total.
 */
fun parseAmount(input: String): Double? {
    val sb = StringBuilder()
    for (c in input) {
        when {
            c in '۰'..'۹' -> sb.append('0' + (c - '۰'))
            c in '٠'..'٩' -> sb.append('0' + (c - '٠'))
            c.isDigit() -> sb.append(c)
            c == '.' || c == '٫' -> sb.append('.')
            // thousands separators and stray spaces are simply ignored
            c == ',' || c == '،' || c == '٬' || c == ' ' || c == '‏' -> Unit
            else -> return null
        }
    }
    val s = sb.toString()
    if (s.isEmpty() || s == ".") return null
    return s.toDoubleOrNull()?.takeIf { it >= 0 && it.isFinite() }
}

/**
 * How much of a thing she holds, sized for a list row. Every coin carries six decimals so the
 * edit field can round-trip a dust balance, but six decimals of ten thousand Tether is
 * precision nobody scans — and on a row it is what pushed the rate, the figure she opened the
 * app to check, off the end of the line. Below one unit every decimal still counts, so a
 * fraction of a Bitcoin keeps all of them; the exact figure is one tap away either way.
 *
 * Truncated, never rounded, for the same reason [faCompact] is: never show more than she has.
 */
fun faHeld(amount: Double, dec: Int): String {
    val d = when {
        amount >= 1_000 -> 2
        amount >= 1 -> 4
        else -> dec
    }.coerceAtMost(dec)
    if (d >= dec) return faDecimal(amount, dec)      // nothing to drop; leave it untouched
    // Settled at the asset's own precision before anything is cut. 10709.13 is not exactly
    // representable as a double — truncating the stored value directly takes off a cent that
    // she really has, which is the one direction this function must never round.
    return faDecimal(
        BigDecimal(amount).setScale(dec, RoundingMode.HALF_UP).setScale(d, RoundingMode.DOWN).toDouble(),
        d,
    )
}

/**
 * How one unit is worth, sized to the number: a dollar reads better in full ("۱۸۷٬۰۰۰"),
 * a gold coin does not ("۱۸۰ میلیون" beats nine digits).
 */
fun faRate(rate: Double): String =
    if (rate >= 1_000_000) faCompact(rate) else faNumber(rate)

/**
 * Wraps a run in Unicode isolates. A latin ticker dropped into Persian text drags the
 * numbers around it out of order — "۴۰ SOL • نرخ ۱۴ میلیون" renders as "SOL ۴۰ • نرخ..."
 * without this. FSI/PDI tells the bidi algorithm to treat the run as one opaque unit.
 */
// Lint flags the isolate characters as spoofing, which is what the check is for: bidi controls
// smuggled into a string change what a reader sees without changing what the code does. Here
// they are the payload rather than a hidden passenger — one FSI, one PDI, balanced, wrapping
// exactly the run named in the argument. Suppressed at the single line that is allowed to
// contain them, so the check keeps working everywhere else.
@Suppress("BidiSpoofing")
fun bidi(s: String): String = "⁨$s⁩"

/**
 * The same isolate, but forced left-to-right, for the digits of a signed amount.
 *
 * [bidi] resolves direction from the first strong character, so an isolate wrapped around
 * "+۱۶۵٫۱ میلیون" takes its direction from the «م» of the magnitude, turns the whole run
 * right-to-left, and throws the sign to the far side of the figure.
 */
// Same reasoning as above: LRI and PDI, balanced, wrapping exactly the argument.
@Suppress("BidiSpoofing")
fun ltrFigure(s: String): String = "⁦$s⁩"

/**
 * «−۴۰۰ میلیون»: signed, compact, and laid out the way Persian actually reads it.
 *
 * Two things have to be true at once, and only one arrangement gives both. The sign belongs on
 * the left of the digits, the way every bank writes it — and the magnitude and «تومان» after it
 * belong at the end of the phrase, which in a right-to-left line is the left end. So the digits
 * and their sign are isolated as one left-to-right island, and the magnitude word is left
 * outside it to flow on in the RTL line: «تومان میلیون −۴۰۰» on screen, «−۴۰۰ میلیون تومان»
 * read aloud. Isolating the pair together, or forcing the whole line left-to-right, each
 * satisfies one of the two and breaks the other.
 */
fun faSignedCompact(toman: Double, positive: Boolean): String =
    faSignedParts(toman, positive).let { (digits, magnitude) ->
        if (magnitude == null) digits else "$digits $magnitude"
    }

/**
 * The same figure, split where a display-sized one needs to set the two halves differently.
 *
 * At 50sp with a magnitude word the same size beside it, the sign sits between two equal runs
 * and stops reading as a sign at all — «۴۰۰ − میلیون» looks like a dash joining two words. The
 * split lets the caller step the magnitude down so the sign clearly belongs to the digits.
 * Returns the isolated signed digits, and the magnitude word, which is null under a thousand.
 */
fun faSignedParts(toman: Double, positive: Boolean): Pair<String, String?> {
    val parts = faCompact(abs(toman)).split(' ', limit = 2)
    return ltrFigure((if (positive) "+" else "−") + parts[0]) to parts.getOrNull(1)
}

/**
 * Plain ASCII digits for the *raw* text-field state — Locale.US on purpose, so the stored
 * string always round-trips through [parseAmount] no matter what the device locale is.
 * Grouping and Persian digits are applied on top of this purely for display.
 */
fun trimNumber(value: Double, dec: Int): String {
    val s = String.format(Locale.US, "%.${dec}f", value)
    return if (s.contains('.')) s.trimEnd('0').trimEnd('.') else s
}

/**
 * Digits grouped for reading while she types, so "85000000" is never a column of zeroes she
 * has to count. Carries the index maps the text field needs to keep the caret in the right
 * place across the separators we inject.
 */
class GroupedDigits(val text: String, val origToDisp: IntArray, val dispToOrig: IntArray)

private fun persianDigit(c: Char): Char? = when (c) {
    in '0'..'9' -> '۰' + (c - '0')
    in '۰'..'۹' -> c
    in '٠'..'٩' -> '۰' + (c - '٠')
    else -> null
}

fun groupDigits(raw: String): GroupedDigits {
    val isSep = { c: Char -> c == '.' || c == '٫' }
    // Only the integer part gets separators; everything after the decimal point is left alone.
    val decimalAt = raw.indexOfFirst(isSep).let { if (it < 0) raw.length else it }
    val intDigits = (0 until decimalAt).count { persianDigit(raw[it]) != null }

    val out = StringBuilder()
    val owner = ArrayList<Int>()          // output index -> original index
    val origToDisp = IntArray(raw.length + 1)

    var seen = 0
    for (i in raw.indices) {
        origToDisp[i] = out.length
        val c = raw[i]
        val digit = persianDigit(c)

        when {
            digit != null -> {
                out.append(digit); owner.add(i)
                if (i < decimalAt) {
                    seen++
                    val remaining = intDigits - seen
                    // Group from the right, never trailing.
                    if (remaining > 0 && remaining % 3 == 0) {
                        out.append('٬'); owner.add(i + 1)
                    }
                }
            }
            isSep(c) -> { out.append('٫'); owner.add(i) }
            else -> { out.append(c); owner.add(i) }
        }
    }
    origToDisp[raw.length] = out.length

    val dispToOrig = IntArray(out.length + 1)
    for (j in 0 until out.length) dispToOrig[j] = owner[j]
    dispToOrig[out.length] = raw.length

    return GroupedDigits(out.toString(), origToDisp, dispToOrig)
}

/** "۵ دقیقه پیش" — vague on purpose; the exact second never matters here. */
fun faAgo(epochMillis: Long, now: Long): String {
    if (epochMillis <= 0) return "هنوز به‌روز نشده"
    val mins = ((now - epochMillis) / 60_000L).coerceAtLeast(0)
    return when {
        mins < 1 -> "همین الان"
        mins < 60 -> "${faNumber(mins.toDouble())} دقیقه پیش"
        mins < 60 * 24 -> "${faNumber((mins / 60).toDouble())} ساعت پیش"
        else -> "${faNumber((mins / (60 * 24)).toDouble())} روز پیش"
    }
}

File: pwa/src/format.ts
/**
 * Format.kt and the date formatters from Timeline.kt, one to one. The policy that matters is
 * the app's, not the browser's: every displayed figure is settled at nine decimals — which only
 * ever erases binary noise — and then *cut*, never rounded, because the number on screen must
 * never be larger than the real one. `toFixed` is the exact-decimal half-up the phone's
 * BigDecimal does, and the cut is made on its string, so no double arithmetic can nudge it.
 */
import { jalaliOf, tehranDay, tehranDayStart, weekStart } from './jalali';

const SETTLE_SCALE = 9;
const PERSIAN = '۰۱۲۳۴۵۶۷۸۹';
const GROUP = '٬';
const DECIMAL = '٫';

export function faDigits(s: string): string {
  return s.replace(/[0-9]/g, (d) => PERSIAN[Number(d)]);
}

/** Settled half-up at nine places, then truncated to `dec`: a plain decimal string, sign kept. */
function settleDown(value: number, dec: number): string {
  const [int, frac = ''] = Math.abs(value).toFixed(SETTLE_SCALE).split('.');
  const cut = dec > 0 ? `${int}.${frac.slice(0, dec)}` : int;
  return value < 0 && /[1-9]/.test(cut) ? `-${cut}` : cut;
}

/** «۱۲٬۳۴۵٫۶۷» from a plain decimal string, with the phone's own negative form. */
function faPlain(decimal: string, keep: 'all' | 'trim' = 'trim'): string {
  const negative = decimal.startsWith('-');
  const [int, frac0 = ''] = decimal.replace('-', '').split('.');
  const frac = keep === 'trim' ? frac0.replace(/0+$/, '') : frac0;
  const grouped = int.replace(/^0+(?=\d)/, '').replace(/\B(?=(\d{3})+(?!\d))/g, ',');
  const body = faDigits(grouped).replaceAll(',', GROUP) + (frac ? DECIMAL + faDigits(frac) : '');
  // What fa-IR's NumberFormat prints for a negative: an LRM, then U+2212.
  return negative ? `‎−${body}` : body;
}

/** Grouped Persian digits, truncated: 118500 → «۱۱۸٬۵۰۰». */
export function faNumber(value: number): string {
  if (!Number.isFinite(value)) return faPlain(String(Math.round(value)));
  if (Math.abs(value) >= 1e21) return faPlain(BigInt(Math.trunc(value)).toString());
  return faPlain(settleDown(value, 0));
}

/**
 * Persian digits with decimals kept: 0.0345 → «۰٫۰۳۴۵». Rounds half-even like the phone's
 * NumberFormat — every money path hands it an already-truncated figure, so this only matters for
 * a caller asking for fewer places than it has.
 */
export function faDecimal(value: number, decimals: number, pad = false): string {
  const factor = 10 ** decimals;
  const scaled = value * factor;
  const floor = Math.floor(scaled);
  const diff = scaled - floor;
  const even = Math.abs(diff - 0.5) < 1e-9 ? (floor % 2 === 0 ? floor : floor + 1) : Math.round(scaled);
  const fixed = decimals > 0 ? (even / factor).toFixed(decimals) : String(even);
  return faPlain(fixed === '-0' ? '0' : fixed.replace(/^-(0(\.0+)?)$/, '$1'), pad ? 'all' : 'trim');
}

export const tomanOf = (rial: number): number => rial / 10;

/** «۳۰۰ میلیون» — the magnitude said out loud, truncated. */
export function faCompact(toman: number, dec = 1, pad = false): string {
  const n = Math.abs(toman);
  let unit: string; let div: number;
  if (n >= 1e12) { unit = 'همت'; div = 1e12; }
  else if (n >= 1e9) { unit = 'میلیارد'; div = 1e9; }
  else if (n >= 1e6) { unit = 'میلیون'; div = 1e6; }
  else if (n >= 1e3) { unit = 'هزار'; div = 1e3; }
  else return faNumber(toman);
  const q = toman / div;
  if (!Number.isFinite(q)) return `${faNumber(q)} ${unit}`;
  const cut = settleDown(q, dec);
  const whole = !/\.\d*[1-9]/.test(cut);
  return `${faPlain(cut, pad && !whole ? 'all' : 'trim')} ${unit}`;
}

const ONES = ['', 'یک', 'دو', 'سه', 'چهار', 'پنج', 'شش', 'هفت', 'هشت', 'نه'];
const TEENS = ['ده', 'یازده', 'دوازده', 'سیزده', 'چهارده', 'پانزده', 'شانزده', 'هفده', 'هجده', 'نوزده'];
const TENS = ['', '', 'بیست', 'سی', 'چهل', 'پنجاه', 'شصت', 'هفتاد', 'هشتاد', 'نود'];
const HUNDREDS = ['', 'صد', 'دویست', 'سیصد', 'چهارصد', 'پانصد', 'ششصد', 'هفتصد', 'هشتصد', 'نهصد'];
const SCALES = ['', 'هزار', 'میلیون', 'میلیارد', 'هزار میلیارد'];

function tripleToWords(n: number): string {
  const parts: string[] = [];
  const h = Math.trunc(n / 100); const r = n % 100;
  if (h > 0) parts.push(HUNDREDS[h]);
  if (r >= 10 && r <= 19) parts.push(TEENS[r - 10]);
  else {
    if (Math.trunc(r / 10) > 0) parts.push(TENS[Math.trunc(r / 10)]);
    if (r % 10 > 0) parts.push(ONES[r % 10]);
  }
  return parts.join(' و ');
}

/** 10_800_000 → «ده میلیون و هشتصد هزار». Exact, never rounded. */
export function faWords(value: number): string {
  if (value === 0) return 'صفر';
  if (value < 0) return `منفی ${faWords(-value)}`;
  const groups: number[] = [];
  let v = BigInt(Math.trunc(value));
  while (v > 0n) { groups.push(Number(v % 1000n)); v /= 1000n; }
  if (groups.length > SCALES.length) return '';
  const parts: string[] = [];
  for (let i = groups.length - 1; i >= 0; i--) {
    const g = groups[i];
    if (g === 0) continue;
    const words = g === 1 && i === 1 ? '' : tripleToWords(g);
    parts.push([words, SCALES[i]].filter((s) => s.trim()).join(' '));
  }
  return parts.join(' و ');
}

export function faWordsToman(value: number): string | null {
  if (value <= 0 || value >= 1e15) return null;
  const words = faWords(Math.round(value));
  return words.trim() ? `${words} تومان` : null;
}

/** Persian, Arabic-Indic or Latin digits, ٫ or . for a decimal, separators ignored; null on anything else. */
export function parseAmount(input: string): number | null {
  let s = '';
  for (const c of input) {
    const code = c.charCodeAt(0);
    if (code >= 0x6f0 && code <= 0x6f9) s += String(code - 0x6f0);
    else if (code >= 0x660 && code <= 0x669) s += String(code - 0x660);
    else if (c >= '0' && c <= '9') s += c;
    else if (c === '.' || c === '٫') s += '.';
    else if (c === ',' || c === '،' || c === '٬' || c === ' ' || c === '‏') continue;
    else return null;
  }
  if (!s || s === '.' || (s.match(/\./g)?.length ?? 0) > 1) return null;
  const n = Number(s);
  return Number.isFinite(n) && n >= 0 ? n : null;
}

/** How much of a thing she holds, sized for a row; truncated at the asset's own precision. */
export function faHeld(amount: number, dec: number): string {
  const d = Math.min(amount >= 1_000 ? 2 : amount >= 1 ? 4 : dec, dec);
  if (d >= dec) return faDecimal(amount, dec);
  const [int, frac = ''] = Math.abs(amount).toFixed(dec).split('.');
  const cut = d > 0 ? `${int}.${frac.slice(0, d)}` : int;
  return faPlain(amount < 0 ? `-${cut}` : cut);
}

export function faRate(rate: number): string {
  return rate >= 1_000_000 ? faCompact(rate) : faNumber(rate);
}

/** Dollars, or nothing: absent whenever it cannot be stated honestly. */
export function usdOf(toman: number, rate: number | null | undefined): number | null {
  return rate != null && rate > 0 && Number.isFinite(rate) && toman > 0 ? toman / rate : null;
}

/** First-strong isolate around a run (FSI … PDI). */
export const bidi = (s: string): string => `⁨${s}⁩`;
/** Left-to-right isolate (LRI … PDI), for a sign and its digits. */
export const ltrFigure = (s: string): string => `⁦${s}⁩`;

export function faSignedParts(toman: number, positive: boolean): [string, string | null] {
  const text = faCompact(Math.abs(toman));
  const at = text.indexOf(' ');
  const digits = at < 0 ? text : text.slice(0, at);
  const magnitude = at < 0 ? null : text.slice(at + 1);
  return [ltrFigure((positive ? '+' : '−') + digits), magnitude];
}
export function faSignedCompact(toman: number, positive: boolean): string {
  const [digits, magnitude] = faSignedParts(toman, positive);
  return magnitude == null ? digits : `${digits} ${magnitude}`;
}

/** ASCII digits for a field's raw state, trailing zeros trimmed. */
export function trimNumber(value: number, dec: number): string {
  const s = value.toFixed(dec);
  return s.includes('.') ? s.replace(/0+$/, '').replace(/\.$/, '') : s;
}

/** Persian digits grouped from the right while she types; decimals left alone. */
export function groupDigits(raw: string): string {
  const persian = (c: string): string | null => {
    const code = c.charCodeAt(0);
    if (c >= '0' && c <= '9') return PERSIAN[code - 48];
    if (code >= 0x6f0 && code <= 0x6f9) return c;
    if (code >= 0x660 && code <= 0x669) return PERSIAN[code - 0x660];
    return null;
  };
  const sep = (c: string) => c === '.' || c === '٫';
  const chars = [...raw];
  let decimalAt = chars.findIndex(sep);
  if (decimalAt < 0) decimalAt = chars.length;
  const intDigits = chars.slice(0, decimalAt).filter((c) => persian(c) != null).length;
  let out = ''; let seen = 0;
  chars.forEach((c, i) => {
    const digit = persian(c);
    if (digit != null) {
      out += digit;
      if (i < decimalAt) {
        seen++;
        const remaining = intDigits - seen;
        if (remaining > 0 && remaining % 3 === 0) out += GROUP;
      }
    } else if (sep(c)) out += DECIMAL;
    else out += c;
  });
  return out;
}

/** «۵ دقیقه پیش» — vague on purpose. */
export function faAgo(epochMillis: number, now: number): string {
  if (epochMillis <= 0) return 'هنوز به‌روز نشده';
  const mins = Math.max(0, Math.trunc((now - epochMillis) / 60_000));
  if (mins < 1) return 'همین الان';
  if (mins < 60) return `${faNumber(mins)} دقیقه پیش`;
  if (mins < 60 * 24) return `${faNumber(Math.trunc(mins / 60))} ساعت پیش`;
  return `${faNumber(Math.trunc(mins / (60 * 24)))} روز پیش`;
}

// ---- dates (Timeline.kt) -------------------------------------------------------------------

export const MONTHS = ['فروردین', 'اردیبهشت', 'خرداد', 'تیر', 'مرداد', 'شهریور', 'مهر', 'آبان', 'آذر', 'دی', 'بهمن', 'اسفند'];
export const WEEKDAYS = ['شنبه', 'یک‌شنبه', 'دوشنبه', 'سه‌شنبه', 'چهارشنبه', 'پنج‌شنبه', 'جمعه'];

export const today = (): number => tehranDay(Date.now());
export const faWeekday = (day: number): string => WEEKDAYS[day - weekStart(day)];
export function faDate(day: number): string {
  const it = jalaliOf(day);
  return `${faNumber(it.day)} ${MONTHS[it.month - 1]} ${faDigits(String(it.year))}`;
}
export const faWeekdayDate = (day: number): string => `${faWeekday(day)} ${faDate(day)}`;
export function faDay(day: number, now = today()): string {
  if (day === now) return 'امروز';
  if (day === now - 1) return 'دیروز';
  return faWeekdayDate(day);
}
export function faClock(epochMillis: number): string {
  const since = epochMillis - tehranDayStart(tehranDay(epochMillis));
  const hour = Math.trunc(since / 3_600_000);
  const minute = Math.trunc(since / 60_000) % 60;
  return faDigits(`${String(hour).padStart(2, '0')}:${String(minute).padStart(2, '0')}`);
}
export function faMoment(at: number, day: number): string {
  return at === tehranDayStart(day) ? faWeekdayDate(day) : `${faWeekdayDate(day)}، ${bidi(faClock(at))}`;
}
export function faDayMoment(at: number, day: number, now = today()): string {
  return at === tehranDayStart(day) ? faDay(day, now) : `${faDay(day, now)}، ${bidi(faClock(at))}`;
}
